mmdet/models/dense_heads/autoassign_head.py

```diff
@@ -0,0 +1,475 @@
+        pos_loss_list, = multi_apply(self.get_pos_loss_single, cls_scores,
+                                     objectnesses, reg_loss_list, gt_labels,
+                                     center_prior_weight_list)
+        pos_avg_factor = reduce_mean(bbox_pred.new_tensor(all_num_gt))
+        pos_loss = sum(pos_loss_list) / pos_avg_factor
+        neg_loss_list, = multi_apply(self.get_neg_loss_single, cls_scores,
+                                     objectnesses, gt_labels, ious_list,
+                                     inside_gt_bbox_mask_list)
+
+        neg_avg_factor = sum(item.data.sum()
+                             for item in center_prior_weight_list)
+        neg_avg_factor = reduce_mean(neg_avg_factor)
+        neg_loss = sum(neg_loss_list) / max(neg_avg_factor, 1)
+
+        center_loss = []
+        for i in range(len(img_metas)):
+            center_loss.append(
+                len(gt_bboxes[i]) /
+                center_prior_weight_list[i].sum().clamp_(min=EPS))
+        center_loss = torch.stack(center_loss).mean() * self.center_loss_weight
+
+        if all_num_gt == 0:
+            pos_loss = bbox_preds[0].sum() * 0
+            dummy_center_prior_loss = self.center_prior.mean.sum(
+            ) * 0 + self.center_prior.sigma.sum() * 0
+            center_loss = objectnesses[0].sum() * 0 + dummy_center_prior_loss
+
+        return dict(
+            loss_pos=pos_loss, loss_neg=neg_loss, loss_center=center_loss)
+
+    def get_targets(self, points, gt_bboxes_list):
+        """Compute regression targets and each point inside or outside gt_bbox
+        in multiple images.
+
+        Args:
+            points (list[Tensor]): Points of each fpn level, each has shape
+                (num_points, 2).
+            gt_bboxes_list (list[Tensor]): Ground truth bboxes of each image,
+                each has shape (num_gt, 4).
+
+        Returns:
+            tuple(list[Tensor]):
+
+                - inside_gt_bbox_mask_list (list[Tensor]): Each
+                  Tensor is with bool type and shape of
+                  (num_points, num_gt), each value
+                  is used to mark whether this point falls
+                  within a certain gt.
+                - concat_lvl_bbox_targets (list[Tensor]): BBox
+                  targets of each level. Each tensor has shape
+                  (num_points, num_gt, 4).
+        """
+
+        num_levels = len(points)
+        concat_points = torch.cat(points, dim=0)
+        # the number of points per img, per lvl
+        num_points = [center.size(0) for center in points]
+        inside_gt_bbox_mask_list, bbox_targets_list = multi_apply(
+            self._get_target_single,
+            gt_bboxes_list,
+            points=concat_points,
+        )
+        bbox_targets_list = [
+            list(bbox_targets.split(num_points, 0))
+            for bbox_targets in bbox_targets_list
+        ]
+        if self.norm_on_bbox:
```

Comment:
Can we do this in _get_target_single to simplify the logics?

Reply:
In _get_target_single，tensor has a shape (num_points_all_levels, 4), we have to do this after line 426 which split the tensor to each level.